x/ibc-rate-limit/ibc_middleware_test.go

```diff
@@ -0,0 +1,747 @@
+	suite.ConfigureTransferChannel()
+	// Sends denom=ibc/C053D637CCA2A2BA030E2C5EE1B28A16F71CCB0E45E8BE52766DC1B241B77878 from A->B.
+	// Rate limit receives "transfer/channel-0/stake" in the packet (because transfer.relay.SendTransfer is called before the middleware)
+	// and should hash it before calculating the value
+	suite.fullSendTest(false)
+}
+
+// Test rate limits are reset when the specified time period has passed
+func (suite *MiddlewareTestSuite) TestSendTransferReset() {
+	suite.ConfigureTransferChannel()
+	// Same test as above, but the quotas get reset after time passes
+	attrs := suite.fullSendTest(true)
+	parts := strings.Split(attrs["weekly_period_end"], ".") // Splitting timestamp into secs and nanos
+	secs, err := strconv.ParseInt(parts[0], 10, 64)
+	suite.Require().NoError(err)
+	nanos, err := strconv.ParseInt(parts[1], 10, 64)
+	suite.Require().NoError(err)
+	resetTime := time.Unix(secs, nanos)
+
+	// Move chainA forward one block
+	suite.ChainA.NextBlock()
+
+	// Reset time + one second
+	oneSecAfterReset := resetTime.Add(time.Second)
+	suite.Coordinator.IncrementTimeBy(oneSecAfterReset.Sub(suite.Coordinator.CurrentTime))
+
+	// Sending should succeed again
+	_, err = suite.AssertSend(true, suite.MessageFromAToB(sdk.DefaultBondDenom, sdkmath.NewInt(1)))
+	suite.Require().NoError(err)
+}
+
+// Test rate while having 2 limits (daily & weekly).
+// Daily is hit, wait until 'day' ends. Do this twice. Second iteration will hit the 'weekly' quota.
+// Then we check that weekly rate limit still hits even after daily quota is refreshed.
+func (suite *MiddlewareTestSuite) TestSendTransferDailyReset() {
+	suite.ConfigureTransferChannel()
+	quotaPercentage := 4
+	suite.initializeEscrow()
+	// Get the denom and amount to send
+	denom := sdk.DefaultBondDenom
+
+	app := suite.GetNeutronZoneApp(suite.ChainA)
+
+	// This is the first one. Inside the tests. It works as expected.
+	channelValue := CalculateChannelValue(suite.ChainA.GetContext(), denom, app.BankKeeper)
+
+	// The amount to be sent is 2% (weekly quota is 4%, daily is 2%)
+	quota := channelValue.QuoRaw(int64(100 / quotaPercentage))
+	sendAmount := quota.QuoRaw(2)
+
+	// Setup contract
+	testOwner := sdk.MustAccAddressFromBech32(testutil.TestOwnerAddress)
+	suite.StoreTestCode(suite.ChainA.GetContext(), testOwner, "./bytecode/rate_limiter.wasm")
+	quotas := suite.BuildChannelQuotaWith2Quotas("weekly", channelTest, denom, 604800, "daily", 4, 4, 86400, 2, 2)
+	addr := suite.InstantiateRLContract(quotas)
+	suite.RegisterRateLimitingContract(addr)
+
+	// send 2% (daily quota is 2%)
+	r, err := suite.AssertSend(true, suite.MessageFromAToB(denom, sendAmount))
+	suite.Require().NoError(err)
+
+	// Calculate remaining allowance in the quota
+	attrs := suite.ExtractAttributes(suite.FindEvent(r.GetEvents(), "wasm"))
+
+	used, ok := sdkmath.NewIntFromString(attrs["daily_used_out"])
+	suite.Require().True(ok)
+
+	suite.Require().Equal(used, sendAmount)
```

Comment:
would be very nice to see such checks after each Send try and after quotas reset to be assured that daily and weekly quotas are as we expect them to be. could you please add such assertions? e.g. check that daily and weekly used out quotas equal to `sendAmount` after the first send, then, once daily is reset, check that the daile quota is reset indeed, and weekly is still at `sendAmount`, and so on for all milestones when we expect the quota usages to change